04-onion-routing.md

```diff
@@ -261,6 +268,197 @@ The reader:
+    - MUST fail an HTLC in set if its `cltv_expiry` elapses.
+    - MAY fail all HTLCs in the set after a reasonable timeout.
+
+### Atomic Multi-path Payment Derivation
+
+Let the _root seed_ `r` be a random 32-byte value. A unique _child preimage_ and
+_child hash_ can be derived for a given `child_index` using the `amp_child`
+function:
+
+```golang
+func amp_child(root_seed [32]byte, child_index uint16) ([32]byte, [32]byte) {
+        preimage := SHA256(root_seed || child_index)
+        hash := SHA256(preimage)
+        return preimage, hash
+}
+```
+where `child_index` is serialized using big-endian byte order.
+
+The sender will use `amp_child` to derive a child hash for each HTLC it sends
+out, and includes the `child_index` used in the derivation in the final hop's
+payload. The receiver will use `amp_child` to settle each HTLC with its
+corresponding child preimage, and also to verify that the correct child hash was
+set by the sender.
+
+In order to provide cryptographic atomicity over the fulfillment of an HTLC set,
+each partial payment `i` also transmits a 32-byte `share`.  Each share `s_i`
+represents an n-of-n secret sharing of `r`, such that:
+
+```
+    r = s_1 ^ ... ^ s_n
+```
+
+If `n` is known upfront, satisfying this equation can be done simply by
+generating all `s_i` randomly.
+
+Otherwise, the sender can generate the shares _adaptively_ by first generating a
+random `r`. For all but the last outgoing HTLC, a random `s_i` is generated and
+included directly. The final HTLC then computes `s_n` as the xor of all other
+shares and `r`:
+
+```
+    s_n = s_1 ^ ... ^ s_n-1 ^ r
+```
+
+If a partial payment fails, this process can be applied recursively to generate
+smaller partial payments, at the same time guaranteeing that the xor of all
+shares results in `r`.
+
+This construction prevents the receiver from learning `r` until all `s_i` have
+arrived. If `r` is successfully reconstructed, the receiver can verify the
+correctness of child hashes used in the HTLC set, and settle use the child
+preimages if they were offered correctly.
+
+The diagram below depicts the relationship between shares, the root seed, child
+preimages, and child hashes in the non-adaptive case. Lowercase variables are
+used to signal independent variables, while capital letters are used to describe
+dependent variables. All independent variables are chosen upfront by the sender.
+```
+    s_1  s_2  s_3    SHARES OF ROOT SEED
+                   
+     |    |    |     
+     └──┐ | ┌──┘       R = s_1 ^ s_2 ^ s_3
+        V V V        
+                   
+          R          ROOT SEED
+                   
+        | | |        
+     ┌──┘ | └──┐       P_i = SHA256(R || child_index_i)
+     V    V    V     
+                   
+    P_1  P_2  P_3    CHILD PREIMAGES
+                   
+     |    |    |     
+     |    |    |       H_i = SHA256(P_i)
+     V    V    V     
+                   
+    H_1  H_2  H_3    CHILD HASHES
+```
+
+### Rationale
```

Comment:
Should we also briefly outline the re-occurring payment use case? Or do we consider this to be more of an application level protocol ,which should be defined in a distinct BIP?

Reply:
I think it would be a good idea to briefly mention it (that would help application developers pick it up).